Add `--read-only` for `service create` and `service update`

yongtang opened the pull request:
**- What I did**

This fix tries to address the issue raised in #29972 where it was not possible to specify `--read-only` for `docker service create` and `docker service update`, in order to have the container's root file system to be read only.

**- How I did it**

This fix adds `--read-only` and update the `ReadonlyRootfs` in `HostConfig` through `service create` and `service update`.

Related docs has been updated.

**- How to verify it**

Integration test has been added.

**- Description for the changelog**

Add `--read-only` for `service create` and `service update`

**- A picture of a cute animal (not mandatory but encouraged)**

![baby-animals-cat-cats-cute-kitten-kittens-desktop-image](https://cloud.githubusercontent.com/assets/6932348/21953561/85c7b6ea-d9ef-11e6-8933-f185d11d2396.jpg)


This fix fixes #29972.

Related SwarmKit PR is https://github.com/docker/swarmkit/pull/1872

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add `--read-only` for `service create` and `service update`

This fix tries to address the issue raised in 29972 where
it was not possible to specify `--read-only` for `docker service create`
and `docker service update`, in order to have the container's root file
system to be read only.

This fix adds `--read-only` and update the `ReadonlyRootfs` in `HostConfig`
through `service create` and `service update`.

Related docs has been updated.

Integration test has been added.

This fix fixes 29972.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

thaJeztah commented on integration-cli/docker_cli_swarm_test.go:
```diff
@@ -1630,3 +1630,35 @@ func (s *DockerSwarmSuite) TestSwarmInitWithDrain(c *check.C) {
+	out, err = d.Cmd("inspect", "--type", "container", "--format", "{{.HostConfig.ReadonlyRootfs}}", containers[0])
+	c.Assert(err, checker.IsNil, check.Commentf(out))
+
+	expected := "touch: /file: Read-only file system"
```
I don't think we need to test this here; this functionality is already tested in [`testReadOnlyFile()`](https://github.com/docker/docker/blob/9af5d7c34039d5e9d8a7dca6ea148147dc482dee/integration-cli/docker_cli_run_test.go#L2813-L2834)

yongtang replied:
Thanks @thaJeztah. That test has been removed.

thaJeztah commented on integration-cli/docker_cli_swarm_test.go:
```diff
@@ -1630,3 +1630,35 @@ func (s *DockerSwarmSuite) TestSwarmInitWithDrain(c *check.C) {
+	c.Assert(err, checker.NotNil, check.Commentf(out))
+	c.Assert(out, checker.Contains, expected)
+
+	out, err = d.Cmd("service", "update", "--read-only=false", "top")
```
I think it would be good to test that running an `docker service update` _without_ specifying the `--read-only` flag,  does not change `ReadonlyRootfs` from `true` to `false`

yongtang replied:
Thanks. I added a unit test to cover this scenario.

thaJeztah commented on integration-cli/docker_cli_swarm_test.go:
```diff
@@ -1630,3 +1630,35 @@ func (s *DockerSwarmSuite) TestSwarmInitWithDrain(c *check.C) {
+	out, err = d.Cmd("service", "update", "--read-only=false", "top")
+	c.Assert(err, checker.IsNil, check.Commentf(out))
+
+	out, err = d.Cmd("service", "inspect", "--format", "{{ .Spec.TaskTemplate.ContainerSpec.ReadonlyRootfs }}", "top")
```
Looking at all these tests; I'm not sure how much of this is tested in SwarmKit, but possibly it could even be changed to a unit test (i.e., test that the client is setting / updating the `ReadonlyRootfs` in the TaskTemplate)

/cc @vdemeester perhaps you have some thoughts on that?

yongtang replied:
Thanks. Some of the test has been changed to unit tests.

thaJeztah reviewed (changes requested):
design looks good to me; left some general thoughts on the tests

thaJeztah commented on cli/command/service/opts.go:
```diff
@@ -479,6 +481,7 @@ func addServiceFlags(cmd *cobra.Command, opts *serviceOptions) {
 	flags.BoolVar(&opts.healthcheck.noHealthcheck, flagNoHealthcheck, false, "Disable any container-specified HEALTHCHECK")
 
 	flags.BoolVarP(&opts.tty, flagTTY, "t", false, "Allocate a pseudo-TTY")
+	flags.BoolVar(&opts.readonlyRootfs, flagReadonlyRootfs, false, "Mount the container's root filesystem as read only")
```
Oh, can you version this flag, so that it only shows if the daemon is docker 1.14 or above?

```go
flags.SetAnnotation(flagReadonlyRootfs, "version", []string{"1.26"})
```

(hm, thinking of which, I think there's some other flags that we need to update as well)

yongtang replied:
Done.

LK4D4 commented:
@yongtang there are some comments from @thaJeztah 
Thanks!

yongtang commented:
@LK4D4 Thanks for the reminder. docker/swarmkit#1872 was merged yesterday. I will work on updating this PR shortly.

yongtang commented:
@thaJeztah Thanks for the review. The PR has been updated. And some of the tests have been changed to unit tests. Please take a look and let me know if there are any issues.

LK4D4 commented:
@yongtang I'm not sure, but failures look real enough

yongtang commented:
@LK4D4 Ah the failure is related to the changes in SwarmKit about explicitly disallow network pluginv1 creation in swarm mode (See discussion in docker/swarmkit#1899, docker/swarmkit#1894, and #30332)

The failure is being addressed in PR #30548 now. Please take a look.

LK4D4 commented:
@yongtang cool, thanks!

LK4D4 commented:
@yongtang just merged it.

yongtang commented:
Thanks @LK4D4! This PR has been updated as well now.

thaJeztah reviewed (commented):
changes LGTM perhaps you can also update the completion scripts for bash and zsh as part of this PR? I think bash change should go here; https://github.com/docker/docker/blob/master/contrib/completion/bash/docker#L2833 (but double check :smile:)

yongtang committed:
Update bash and zsh completion for `service create/update --read-only`

This commit updates bash and zsh completion for flag `--read-only`
in `service create/update`.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @thaJeztah. The PR has been updated with related bash and zsh completion added.

thaJeztah reviewed (approved):
LGTM, thanks!

LK4D4 commented:
LGTM

vieux commented:
@yongtang can you please open a PR to update `docs/api/version-history.md`

yongtang commented:
Thanks @vieux for the reminder. The PR #31145 has been opened.
